### ISSUE TITLE
Multiple scene updates followed by a scene load causes a crash

### ISSUE THREAD

**Issue by msmollin** [tangrams/tangram-es#1677]
**TO REPRODUCE THE ISSUE, FOLLOW THESE STEPS:**

- Make sample iOS project using Tangram-es 0.8.1 release
- Load a scene file from disk (did not test from web)
- Apply 3+ scene updates and an async scene load in that order

**RESULT:**

Crash!

**EXPECTED RESULT:**

Not to crash!

**ENVIRONMENT:**

-iOS 10.x both simulator and device.

**OTHER:**

I tried to reproduce this with the debug builds available on ios.mapzen.com but to no avail. It seems the debug builds either take a slightly different code path, OR are just so much slower that tangram can cope with all the scene updates. I've attached the view controller code i used to do it to speed up the process. 
[ViewController.swift.zip](https://github.com/tangrams/tangram-es/files/1354501/ViewController.swift.zip)

**Comment by msmollin** [comment]
Did a little more testing on this. You don't actually need to do a scene load at all. It's just multiple consecutive updates to scene file globals that causes the crash.

**Comment by hjanetzek** [comment]
The issue is probably caused by missing synchronization between sync and async scene loading/update calls. When async scene loading/updates are in progress loadScene must wait for them to finish.

**Comment by tallytalwar** [comment]
Hmm, I am not sure if a synchronous scene load/update is at conflict here from the usage in the ios-sdk. That said that is surely a possibility which we should address.

Also, I have so far not able to reproduce this race in the osx demo app, more soon.

**Comment by tallytalwar** [comment]
Few updates here.

1. I was able to repro "a" crash on ios demo app by using the following:

```
 NSMutableArray<TGSceneUpdate *>* updates = [[NSMutableArray alloc]init];
    [updates addObject:[[TGSceneUpdate alloc]initWithPath:@"global.sdk_mapzen_api_key" value:apiKey]];

   
    [updates addObject:[[TGSceneUpdate alloc]initWithPath:@"global.sdk_mapzen_api_key" value:@"mapzen-iSPPZHx"]];
    [updates addObject:[[TGSceneUpdate alloc]initWithPath:@"global.sdk_path_overlay" value: @"true"]];
    
    NSMutableArray<TGSceneUpdate *>* pathUpdate = [[NSMutableArray alloc]init];
    [pathUpdate addObject:[[TGSceneUpdate alloc]initWithPath:@"global.sdk_path_overlay" value: @"true"]];
    NSString* yaml = @"{ import: [ https://raw.githubusercontent.com/tangrams/walkabout-style/gh-pages/walkabout-style.yaml, https://raw.githubusercontent.com/tangrams/walkabout-style/gh-pages/themes/label-5.yaml ] }";
    
    [super loadSceneFromYAML:yaml relativeToURL:[NSURL URLWithString:@""] withUpdates:updates];
    
    [super updateSceneAsync:pathUpdate];
    [super updateSceneAsync:pathUpdate];
    [super updateSceneAsync:pathUpdate];
    [super updateSceneAsync:pathUpdate];
    
    [super loadSceneFromYAML:yaml relativeToURL:[NSURL URLWithString:@""] withUpdates:updates];
```

However same set of operations do not cause any crashes on android or osx.

Will be checking debug builds on the ios demo app to narrow down this issue!

**Comment by karimnaaji** [comment]
On iOS using your code @tallytalwar I have been seeing only crashes in core assembly with a random pattern on a bad memory access from different threads. After trying on OS X, it appears to be an issue where the two async calls `loadSceneYamlAsync` and `updateSceneAsync`  are directly following each other. 

It's reproduce under the OS X demo application in RELEASE build:

```diff
diff --git a/platforms/common/glfwApp.cpp b/platforms/common/glfwApp.cpp
index 3fa674046..cfe0bfb60 100644
--- a/platforms/common/glfwApp.cpp
+++ b/platforms/common/glfwApp.cpp
@@ -62,11 +62,10 @@ void loadSceneFile(bool setPosition) {
         updates.push_back(SceneUpdate("global.sdk_mapzen_api_key", mapzenApiKey));
     }
 
-    if (!sceneYaml.empty()) {
-        map->loadSceneYamlAsync(sceneYaml, sceneFile, setPosition, updates);
-    } else {
-        map->loadSceneAsync(sceneFile, setPosition, updates);
-    }
+    sceneYaml = "{ import: [ https://raw.githubusercontent.com/tangrams/walkabout-style/gh-pages/walkabout-style.yaml, https://raw.githubusercontent.com/tangrams/walkabout-style/gh-pages/themes/label-5.yaml ] }";
+
+    map->loadSceneYamlAsync(sceneYaml, sceneFile, setPosition, updates);
+    map->updateSceneAsync({SceneUpdate{"global.sdk_path_overlay", "true"}});
 }
 
```

**Comment by tallytalwar** [comment]
I think #1705 fixes the issue this is running into. I have to test this a bit more and will finalize the PR by tomorrow.

**Comment by tallytalwar** [comment]
Going to reopen this as #1722 does not cater to this issue. Should be resolved by #1705 though.
